docs/propose_changelog.py

```diff
@@ -1,177 +1,119 @@
+    "Refactoring": lambda pr: not contains_label(pr, "bug") and contains_label(pr, "refactoring"),
+    "Documentation": lambda pr: contains_label(pr, "documentation"),
+}
+
+# This is the main configuration point: The changelog will have the same tree structure as this nested dict. The leaves
+# however will be replaced with something roughly equivalent to `list(filter(func, PRs))` if `func` is the corresponding
+# function constituting a leave of `CATEGORIES`. In order to save some typing, the bulk of the categorisation is
+# formulated as a cartesion product of `MAIN_CATEGORIES` and `COMPONENTS` but this is only for convenience. Feel free
+# to change this or amend this by hand afterwards.
+CATEGORIES = {
+    "User Input Changes": (lambda pr: contains_label(pr, "component: user input")),
+} | {
+    main_cat: {
+        # This is important: If you create the lambda in-place, variables are captured by reference and the
+        # corresponding value is only fetched upon a call. This leads to all lambdas using the last value of the
+        # iteration.
+        name: make_lambda(main_condition, component)
+        for name, component in COMPONENTS.items()
+    }
+    | {"other": make_lambda(main_condition)}
+    for main_cat, main_condition in MAIN_CATEGORIES.items()
+}
+
+
+def contains_label(issue, label):
+    """Helper function to check if an issue is labelled by label."""
+    return label in map(lambda lab: lab.name, issue.labels)
+
+
+def categorise(prs, categories_or_condition):
+    """Recursively run over the given categories and filter the PRs according to their conditions."""
+    if not isinstance(categories_or_condition, dict):
+        return list(filter(categories_or_condition, prs))
+    return {key: categorise(prs, val) for key, val in categories_or_condition.items()}
+
+
+def apply_to_leaves(func, mapping):
+    """Helper function to recursively apply to the leaves of a nested dictionary (applying to values of a list individually)."""
+    if isinstance(mapping, dict):
+        return {key: apply_to_leaves(func, val) for key, val in mapping.items()}
+    if isinstance(mapping, list):
+        return list(map(func, mapping))
+    return func(mapping)
+
+
+def to_string(categories):
+    """Transform our nested dictionary of GH PR objects into something readable."""
+    return yaml.dump(apply_to_leaves(lambda pr: f"{pr.title} #{pr.number}", categories))
+
+
+def pull_requests(gh_key, version):
+    """Query the Github API for the kind of PRs we need."""
+    return Github(gh_key).search_issues(
+        f'repo:ComputationalRadiationPhysics/picongpu type:pr is:merged milestone:"{version}" label:changelog'
+    )
+
+
+def main(version, gh_key=None):
+    """Main logic: Download, categorise, print."""
+    print(to_string(categorise(pull_requests(gh_key, version), CATEGORIES)))
+
+
+if __name__ == "__main__":
+    main(*sys.argv[1:])
```

Comment:
I think a help would be useful, possible via [`argparse`](https://docs.python.org/3/library/argparse.html)

Reply:
Yeah, I know. But I dislike `argparse` and  prefer [click](https://click.palletsprojects.com). But then again, that's an (almost) unnecessary (non-standard-library) dependency, so for scripts as simple as this one, I typically go with just `sys.argv` directly until a cautious reviewer forces me to use one or the other. Consider it done.